cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+      getNext(dest, dir, dest_dir, &dest_entry, &dest_iter,
+              &dest_st, do_fsck, false, pstats);
+    } else {
+      // A destination entry was added
+      pstats->Lookup(SHRINKWRAP_STAT_DEST_ENTRIES)->Inc();
+    }
+
+    if (!src_entry && !dest_entry) {
+      // If we have visited all entries in both we break
+      break;
+    } else if (!src_entry) {
+      cmp = 1;
+    } else if (!dest_entry) {
+      cmp = -1;
+    } else {
+      cmp = strcmp(src_entry, dest_entry);
+    }
+
+    if (cmp <= 0) {
+      // Compares stats to see if they are equivalent
+      if (cmp == 0 && cvmfs_attr_cmp(src_st, dest_st, dest)
+        // Also check internal hardlink consistency in destination file system
+        // where applicable:
+        && (dest_st->cvm_checksum == NULL
+          || dest->is_hash_consistent(dest->context_, dest_st))) {
+        if (S_ISDIR(src_st->st_mode) && recursive) {
+          add_dir_for_sync(src_entry, recursive);
+        }
+        continue;
+      }
+      // If not equal, bring dest up-to-date
+      switch (src_st->st_mode & S_IFMT) {
+        case S_IFREG:
+          if (!handle_file(src, src_st, dest, dest_st, src_entry, pstats))
+            return false;
+          break;
+        case S_IFDIR:
+          if (!handle_dir(src, src_st, dest, dest_st, src_entry))
+            return false;
+          if (recursive)
+            add_dir_for_sync(src_entry, recursive);
+          break;
+        case S_IFLNK:
+          // Should likely copy the source of the symlink target
+          if (dest->do_symlink(dest->context_, src_entry,
+                               src_st->cvm_symlink, src_st) != 0) {
+            LogCvmfs(kLogCvmfs, kLogStderr,
+              "Traversal failed to symlink %s->%s : %d : %s",
+              src_entry, src_st->cvm_symlink, errno, strerror(errno));
+            return false;
+          }
+          break;
+        default:
+          LogCvmfs(kLogCvmfs, kLogStderr,
+            "Unknown file type for %s : %d", src_entry, src_st->st_mode);
+          return false;
+      }
+    /* Dest contains something missing from Src */
+    } else {
+      switch (dest_st->st_mode & S_IFMT) {
+        case S_IFREG:
+        case S_IFLNK:
+          if (dest->do_unlink(dest->context_, dest_entry) != 0) {
+            LogCvmfs(kLogCvmfs, kLogStderr,
+              "Failed to unlink file %s", dest_entry);
+            return false;
+          }
+          break;
+        case S_IFDIR:
+          Sync(dest_entry, src, dest, true, pstats, false);
+          if (dest->do_rmdir(dest->context_, dest_entry) != 0) {
+            LogCvmfs(kLogCvmfs, kLogStderr,
+              "Failed to remove directory %s", dest_entry);
+            return false;
+          }
+          break;
+        default:
+          // Unknown file type, should print error (what stream? log?)
+          LogCvmfs(kLogCvmfs, kLogStderr,
+            "Unknown file type for %s : %d", dest_entry, dest_st->st_mode);
```

Comment:
Should this be printed out differently so it is more clear what type and more the file is?

Reply:
yes, on it